src/sentry/integrations/pagerduty/integration.py

```diff
@@ -0,0 +1,166 @@
+                "mappedSelectors": {
+                    "service": {"choices": service_options, "placeholder": "Select a service"}
+                },
+                "columnLabels": {"service": "Service"},
+                "mappedColumnLabel": "Sentry Project",
+            }
+        ]
+
+        return fields
+
+    def update_organization_config(self, data):
+
+        if "project_mapping" in data:
+            project_ids_and_services = data.pop("project_mapping")
+            with transaction.atomic():
+                PagerDutyServiceProject.objects.filter(
+                    organization_integration_id=self.org_integration.id
+                ).delete()
+
+                for p_id, s in project_ids_and_services.items():
+                    # create the record in the table
+                    project = Project.objects.get(pk=p_id)
+                    service = self.services[s["service"]]
+                    PagerDutyServiceProject.objects.create(
+                        organization_integration=self.org_integration,
+                        project=project,
+                        service_name=service["name"],
+                        service_id=service["id"],
+                        integration_key=service["integration_key"],
+                    )
+
+    def get_config_data(self):
+        config = self.org_integration.config
+        project_mappings = PagerDutyServiceProject.objects.filter(
+            organization_integration_id=self.org_integration.id
+        )
+        data = {}
+        for pm in project_mappings:
+            data[pm.project_id] = {"service": pm.service_id}
+        config = {}
+        config["project_mapping"] = data
+        return config
+
+    @property
+    def services(self):
+        return self.model.metadata["services"]
+
+
+class PagerDutyIntegrationProvider(IntegrationProvider):
+    key = "pagerduty"
+    name = "PagerDuty"
+    metadata = metadata
+    features = frozenset([IntegrationFeatures.ALERT_RULE])
+    integration_cls = PagerDutyIntegration
+
+    setup_dialog_config = {"width": 600, "height": 900}
+
+    def get_pipeline_views(self):
+        return [PagerDutyInstallationRedirect()]
+
+    def build_integration(self, state):
+        config = json.loads(state.get("config"))
+        account = config["account"]
+        # PagerDuty gives us integration keys for various things, some of which
+        # are not services. For now we only care about services.
+        services = filter(lambda x: x["type"] == "service", config["integration_keys"])
+
+        return {
+            "name": account["name"],
+            "external_id": account["subdomain"],
+            # make it easy to get a service using the format:
+            #   "service_id": {
+            #       "integration_key": <key>.
+            #       "service_name": <name>,
+            #       "service_id": <id>,
+            #   }
+            "metadata": {"services": dict([(s["id"], s) for s in services])},
```

Comment:
@MeredithAnya I feel a bit uncomfortable relying on metadata for core functionality. At the DB level, we have no guarantee that `services` will exist. I would much prefer to have a service table so we can make this more SQL friendly.